hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/MetricsRegionServerWrapperImpl.java

```diff
@@ -275,7 +276,7 @@ public int getFlushQueueSize() {
 
   @Override
   public long getBlockCacheCount() {
-    return this.blockCache.map(BlockCache::getBlockCount).orElse(0L);
+    return this.blockCache != null ? this.blockCache.getBlockCount() : 0L;
```

Comment:
What is the advantage of the changes in this file? It just increases the branch complexity...

Reply:
IMHO, if we use Optional<> for return value only and not for internal computation, that would be good.
In this file, we define blockCache, mobFileCache, cacheStats, l1Stats, l2Stats all as optional but just for computational purpose and not for actually returning them. 
Using null check for computation is fine and better compared to optional comparison as long as we don't return nullable object from any method.
Thoughts?

Reply:
I think the advantage is comparing to Optional.isPresent, not Optional.map? Here we do not acll Optional.isPresent, just call Optional.map and then orElse?

Reply:
Agree @Apache9 However, Optional.map internall does isPresent() check.

Reply:
The intention here is that, the method we use to get these values returns an Optional, so I do not think it is necessary to unwrap them and then explicitly using null checks to do computation... What we should avoid is calling Optional.of before calling a method, which introduces extra object creation and null comparison.

Reply:
Valid and agree with this. Although one thing to notice is initBlockCache() is called only once during initialization and not multiple times. Whereas all these getters can be called multiple times and hence, those many isPresent() checks followed by Optional.ofNullable() calls.
Rather we can just do unwrap once during init() and then have simple null check and no Optional.ofNullable() during every get call. Thought?

Reply:
@Apache9 @petersomogyi I was just checking online and came across this: https://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments
Looks like we have tried to cover all that is mentioned in 1st ans

Reply:
> Agree @Apache9 However, Optional.map internally does isPresent() check.

isPresent is just like the null check, the intention is to avoid explicit isPresent calls. The provided page is about using Optional as parameters, but here it is a field, not a parameter, and do not need to do extra Optional.of to create it, so I do not get the point why we must unwrap it and explicitly then use null check? What is the advantage?

Reply:
Advantages with this patch:

1. we can avoid using Optional field (minor impact since we are not serializing this class)
2. we can unwrap only one time during initialization and then use simple null check. Otherwise Optional.map() anyways does Optional.of for all of these getter methods:
```
    public<U> Optional<U> map(Function<? super T, ? extends U> mapper) {
        Objects.requireNonNull(mapper);
        if (!isPresent())
            return empty();
        else {
            return Optional.ofNullable(mapper.apply(value));
        }
    }
```
Hence, isn't it better to have null check rather than going into this mapper function to apply within Optional.of path?

Reply:
Oh, we are returning primitive types here, Optional will introduce unnecessary boxing/unboxing. Fine. Let's use null check then.